### PR TITLE
Vecenvs

### PR DESCRIPTION
**Pull request by Sharad24**

### CODE REVIEW

**Comment by Sharad24** [comment]
TD3 and DDPG work but looks like there's something wrong with counting the number of episodes.

**Comment by Sharad24** [comment]
To do:

- [ ] DQN
- [ ] SAC
- [ ] PPO
- [ ] VPG
- [ ] This PR, as of now, removes functionality to run only a single environment (i.e., Not Vectored). Ideally we should reshape observations to (-1, obs_shape) when there's only a single env to keep the algorithms agnostic of vectorization/non-vectorization.
- [ ] I think I might be making a mistake somewhere with the squeeze and unsqueeze operations. As pytorch is printing out a warning. Which ideally it shouldn't. I think I might be missing out on a mean operation somewhere.

Lets benchmark at least 3 environments for each algo to see the benefit of this? (Record seeds).
Otherwise, it does not make much sense. 

Feel free push atomic changes on this branch on the above To dos. This way we can also beautify the repo with results later on.

**Comment by Sharad24** [comment]
Did DQN even work with other envs? @Het-Shah

**Comment by Het-Shah** [comment]
> Did DQN even work with other envs? @Het-Shah

I checked only CartPole

**Comment by Sharad24** [comment]
Rollout Buffers is the same as SB3, but I think once we can get this working, we can reduce stuff in there

**Comment by Het-Shah** [comment]
So what are you currently working on here?

**Comment by Sharad24** [comment]
Yeah.. I'm working on off policy algos

**Comment by Het-Shah** [comment]
Okay cool

**Comment by Sharad24** [comment]
Hmm, so the major problem with algos not working is probably due to missing copy operations. Not 100% sure about PPO. Definitely true with SAC, etc

**Comment by Sharad24** [comment]
@Het-Shah Were you able to take a look? SAC works pretty good now (On Pendulum-v0)

**Comment by Sharad24** [comment]
Vanilla DQN works, have to check others

**Comment by Sharad24** [comment]
Others also work now. @Het-Shah Was the reward for any of the DQN's consistently 500 for a long series of episodes? Or hovering in 300s, 400s and sometimes 500?

**Comment by Sharad24** [comment]
Some hardcoded things are there in buffer.py (action dim), which need to be set according to the env.

**Comment by Sharad24** [comment]
Not working on this until evening, so feel free to push

**Comment by Het-Shah** [comment]
> Others also work now. @Het-Shah Was the reward for any of the DQN's consistently 500 for a long series of episodes? Or hovering in 300s, 400s and sometimes 500?

Hovers around in 400 - 500

**Comment by Het-Shah** [comment]
> Some hardcoded things are there in buffer.py (action dim), which need to be set according to the env.

Okay cool will do this.

**Comment by Het-Shah** [comment]
Let me know if anything else is needed.

**Comment by Sharad24** [comment]
Problem with PPO is that the policy networks are never getting updated.

File: genrl/deep/agents/ppo1/ppo1.py [hunk 1]
```diff
@@ -132,151 +137,152 @@ def create_model(self):
...
-        self.value_fn.value_hist = torch.cat(
-            [self.value_fn.value_hist, val.unsqueeze(0)]
-        )
+        return action.detach().cpu().numpy(), val, c_new.log_prob(action)
 
-        return action
+    def evaluate_actions(self, obs, old_actions):
+        val = self.value_fn.get_value(obs)
+        print(val.shape)
+        _, dist = self.policy_old.get_action(obs)
```
Comment (Sharad24) [inline genrl/deep/agents/ppo1/ppo1.py:167]: This should've been: `_, dist = self.policy_new.get_action(obs)`.

@Het-Shah 

cc: @ajaysub110 (Kill me already)

**Comment by Sharad24** [comment]
Reaching 431 rewards on CartPole-v1 in 21 episodes

File: genrl/deep/agents/ppo1/ppo1.py (hunk 1 above)
Comment (Het-Shah) [reply genrl/deep/agents/ppo1/ppo1.py]: Lol. How did you even find this.

Comment (Sharad24) [reply genrl/deep/agents/ppo1/ppo1.py]: Fell down to running python debugger on it.

Comment (Het-Shah) [reply genrl/deep/agents/ppo1/ppo1.py]: lol nice

**Comment by Sharad24** [comment]
Alright, everything's done now just cleaning up's left.

File: genrl/deep/agents/dqn/dqn.py [hunk 2]
```diff
@@ -341,18 +349,34 @@ def get_td_loss(self) -> torch.Tensor:
...
+#             loss = (q_value.unsqueeze(1) - expected_q_value.detach()).pow(2).mean()
+#         # loss = F.smooth_l1_loss(q_value,expected_q_value)
+# 
+# =======
         if self.categorical_dqn:
             loss = -(Variable(projection_dist) * dist.log()).sum(1).mean()
+# >>>>>>> master
         else:
             if self.prioritized_replay:
                 loss = (q_value - expected_q_value.detach()).pow(2) * weights
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/dqn.py:382]: @Het-Shah Can you check this? I wasnt sure how to resolve the conflict.

(Confused between the many if statements)

Comment (Het-Shah) [reply genrl/deep/agents/dqn/dqn.py:382]: Yea sure.

**Comment by Sharad24** [comment]
Will resolve build errors and LGTM alerts after a few hours

**Comment by Het-Shah** [comment]
I'll do it

**Comment by Sharad24** [comment]
Oh ok, thanks! 👯

**Comment by Sharad24** [comment]
There might be a lot of build errors actually... so it would be maybe helpful to test locally, I think..

**Comment by Het-Shah** [comment]
I can't really run locally my ubuntu's memory is full so it sucks :(. I'll do it the slow way. Squash and merge this pr anyways.

**Comment by Het-Shah** [comment]
Okay so one direct error that i see is that we cannot use `env=gym.make(...)`. We have to use `env = venv(...)` otherwise it won't work. I have updated all the tests now with this.

**Comment by Het-Shah** [comment]
SAC, TD3, VPG, DDPG work now. PPO, DQN, A2C is left.

**Comment by Het-Shah** [comment]
Also, if I am not wrong A2C doesn't use a buffer right? We are assuming it'll that is why A2C is failing

**Comment by Sharad24** [comment]
Oh I forgot A2C.

**Comment by Het-Shah** [comment]
there are some errors in the save and load thing which will go away after @ajaysub110's pr.

**Comment by Sharad24** [comment]
A2C is very similar in nature to PPO, so it should use a rollout buffer

**Comment by Het-Shah** [comment]
Yea then the implementation of A2C is left that is why there is no rollout buffer in A2C

**Comment by Sharad24** [comment]
Cool, working in it will push it in a jiffy

**Comment by Het-Shah** [comment]
> SAC, TD3, VPG, DDPG work now. PPO, DQN, A2C is left (there is some error in the serialVecEnv have a look at this). Also there is some error in save and load params.

Cool. PPO is done too now, DQN has some error in pushing actions for vecenv they are just integers

**Comment by Sharad24** [comment]
Looks like we're doing A2C in VPG :p

**Comment by Het-Shah** [comment]
OOOF.

**Comment by Sharad24** [comment]
Yeah VPG should not have a critic. Nor should it be calculating Advantage estimates. 

https://github.com/SforAiDl/genrl/blob/b52fcd1226428a2c9844e952e6eb05c40025c879/genrl/deep/agents/vpg/vpg.py#L186

Got overlooked it seems.

**Comment by Sharad24** [comment]
Alright this is done as well, should we correct VPG here or in a separate PR?

**Comment by Het-Shah** [comment]
I guess do it here only. Will be done with working agents.

**Comment by Sharad24** [comment]
Alright will resolve after eating dinner

**Comment by Sharad24** [comment]
Only some last things I guess, but should probably resolve before merging:

1. Algos with only gym.Env or venv but n_envs=1
2. Atari envs (DQN)

**Comment by Het-Shah** [comment]
> Algos with only gym.Env or venv but n_envs=1

Regarding this can we do something like this inside trainer we see if our env is an instance of gym.Env, then we set the env=venv("..", 1)

**Comment by Sharad24** [comment]
DQN's only left right?

**Comment by Het-Shah** [comment]
Yea.

File: genrl/deep/agents/a2c/a2c.py [hunk 3]
```diff
@@ -137,12 +144,19 @@ def create_model(self) -> None:
...
+        )
+
+        # self.traj_reward = []
+        # self.actor_hist = torch.Tensor().to(self.device)
+        # self.critic_hist = torch.Tensor().to(self.device)
 
-        self.actor_loss_hist = torch.Tensor().to(self.device)
-        self.critic_loss_hist = torch.Tensor().to(self.device)
+        # self.actor_loss_hist = torch.Tensor().to(self.device)
+        # self.critic_loss_hist = torch.Tensor().to(self.device)
```
[the remark this answers is not in the document]
Comment (Het-Shah) [reply genrl/deep/agents/a2c/a2c.py:159]: Yea i'll do these in all a2c files

File: genrl/deep/__init__.py [hunk 4]
```diff
@@ -1 +1,22 @@
...
+    get_model,
+    save_params,
+    load_params,
+    venv,
+    SerialVecEnv,
+    SubProcessVecEnv,
+    NormalActionNoise,
+    OrnsteinUhlenbeckActionNoise,
+    Logger,
+    set_seeds,
```
[the remark this answers is not in the document]
Comment (Het-Shah) [reply genrl/deep/__init__.py:18]: I agree with this

**Comment by Sharad24** [comment]
None of the off-policy algos are working 🙃

**Comment by Het-Shah** [comment]
> None of the off-policy algos are working :upside_down_face:

is there any error that is showing up or loss is not calculated properly?

**Comment by Sharad24** [comment]
Its not training.

**Comment by Sharad24** [comment]
Only DQN's not training actually, all others are training

**Comment by Sharad24** [comment]
@Het-Shah Could you take a look once?

File: genrl/deep/agents/dqn/dqn.py [hunk 5]
```diff
@@ -269,20 +270,22 @@ def select_action(self, state: np.ndarray, explore: bool = True) -> np.ndarray:
         :returns: Action based on the state and epsilon value 
         :rtype: int, float, ... 
         """
-        if explore == True:
-            if np.random.rand() <= self.epsilon:
-                return self.env.action_space.sample()
+        # if explore == True:
+        #     if np.random.rand() <= self.epsilon:
+        #         return self.env.action_space.sample()
```
Comment (Het-Shah) [inline genrl/deep/agents/dqn/dqn.py:275]: Why did you comment this out?

Comment (Sharad24) [reply genrl/deep/agents/dqn/dqn.py:275]: Same thing is happening 7/8 lines below

Comment (Het-Shah) [reply genrl/deep/agents/dqn/dqn.py:275]: yea nvm earlier it was different.

**Comment by Het-Shah** [comment]
the shape's and all seems fine to me, there is some silly error that I am missing.

**Comment by Sharad24** [comment]
Not sure. I'm raising two issues arising from here to keep track.

#150 and #151

**Comment by Zeus3101** [comment]
Are you merging this? I'll need to work on DQN for adding Atari env support anyway. I might get something from a fresh perspective as I haven't seen any of the code in this PR yet.

File: genrl/deep/agents/a2c/a2c.py [hunk 6]
```diff
@@ -137,12 +144,12 @@ def create_model(self) -> None:
...
-        self.critic_hist = torch.Tensor().to(self.device)
-
-        self.actor_loss_hist = torch.Tensor().to(self.device)
-        self.critic_loss_hist = torch.Tensor().to(self.device)
+        self.rollout = RolloutBuffer(
+            2048,
+            self.env.observation_space,
+            self.env.action_space,
+            n_envs=self.env.n_envs,
+        )
```
Comment (Zeus3101) [inline genrl/deep/agents/a2c/a2c.py:152]: Is 2048 a standard number or can we make that more generic like replay_size?

Comment (Sharad24) [reply genrl/deep/agents/a2c/a2c.py:152]: Yeah, we can change this.

**Comment by Sharad24** [comment]
@ajaysub110 Can you take a final look? Will merge, if everything's good.

File: genrl/deep/common/VecEnv.py [hunk 7]
```diff
@@ -98,7 +93,7 @@ def seed(self, seed):
         """
         Set seed for reproducability in all environments
```
Comment (ajaysub110) [inline genrl/deep/common/VecEnv.py:94]: Hehe spelling mistake

File: genrl/deep/common/VecEnv.py [hunk 8]
```diff
@@ -178,22 +201,26 @@ def render(self, mode="human"):
...
+        # if mode == "human":
+        #     # make_grid(self.images())
+        #     import cv2  # noqa
+
+        #     cv2.imshow("vecenv", out_image[:, :, ::-1])
+        #     cv2.waitKey(1)
+        # elif mode == "rgb_array":
+        #     return out_image
+        # else:
+        #     raise NotImplementedError
```
Comment (ajaysub110) [inline genrl/deep/common/VecEnv.py:223]: Why are you rendering only `envs[0]`? Were you facing some issues with the grid? If so, should an issue be opened?

File: genrl/deep/common/VecEnv.py [hunk 9]
```diff
@@ -178,22 +201,26 @@ def render(self, mode="human"):
...
 
 
 class SubProcessVecEnv(VecEnv):
```
Comment (ajaysub110) [inline genrl/deep/common/VecEnv.py:226]: Doc line for `SubProcessVecEnv` is the same as `SerialVecEnv`

File: genrl/deep/common/buffers.py [hunk 10]
```diff
@@ -3,9 +3,81 @@
 import random
 import numpy as np
 from typing import Tuple
+from .utils import get_obs_shape, get_action_dim
 
 
 class ReplayBuffer:
+    def __init__(self, size, env):
```
Comment (ajaysub110) [inline genrl/deep/common/buffers.py:10]: Docs for Replay Buffer need to be added. We can do that somewhere else since this seems holding up other PRs

**Comment by Sharad24** [comment]
Done here, merge at will.

File: genrl/deep/common/VecEnv.py (hunk 8 above)
Comment (Sharad24) [reply genrl/deep/common/VecEnv.py:223]: Yeah, I'm not able to render the grid. Not sure what's happening here. Raised an issue. #153

**Comment by ajaysub110** [comment]
Other than these points of clarification, code looks really good. Much cleaner than most other parts of the repo. Good job guys!

File: genrl/deep/common/buffers.py (hunk 10 above)
Comment (Sharad24) [reply genrl/deep/common/buffers.py:10]: Yeah. Raised an issue for that. I'll work on that. #154

File: genrl/deep/common/VecEnv.py (hunk 7 above)
Comment (Sharad24) [reply genrl/deep/common/VecEnv.py]: Resolved

File: genrl/deep/common/VecEnv.py (hunk 9 above)
Comment (Sharad24) [reply genrl/deep/common/VecEnv.py:221]: Changed

**Comment by Sharad24** [comment]
Will merge, if checks pass. Let me know if there's anything else.
